buildah: add a new 0.3 version removing privileged

vdemeester opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Using the `vfs` storage driver in a container is similar to using
`overlay` but has the benefit to not require the `privileged`
securityContext.

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

vdemeester commented on task/buildah/0.3/README.md:
```diff
@@ -0,0 +1,63 @@
+  building images. (_default:_ `""`)
+* **PUSH_EXTRA_ARGS**: Extra parameters passed for the push command when
+  pushing images. (_default:_ `""`)
+* **STORAGE_DRIVER**: Set build storage driver to use (_default:_ `"vfs"`_)
```
I wonder if we should have this parameter at all (and if we should document it or not)

chmouel replied:
Isnt there some drawbacks using vfs ? Being much slower? (maybe i am confused tho which option is the really slow one here)

The user may want to adjust it depend of their build workload and not being bothered about running it as Root

vdemeester replied:
> Isnt there some drawbacks using vfs ? Being much slower? (maybe i am confused tho which option is the really slow one here)

Well, usually there is, but in a container in a cluster, the difference is very very very small (cc @gabemontero). It is small enough that we cane consider using `vfs` by default, and thus allowing us to not required `privileged`.

chmouel replied:
I mean the drawback to don't give the choise the user with a  params.

vdemeester replied:
Ah :sweat_smile: well keeping it means it might or might not work (if the user provides `overlay` for `STORAGE_DRIVER` without being able to run it as `privileged` it will fail)

gabemontero replied:
there have also been enhancements in buildah wrt vfs since this example was first constructed

some recent experimentation on our part in OCP showed the difference now was not discernable.

IIRC I believe historically this example was using 1.11

but I see for example @vdemeester 's update moves to `quay.io/buildah/stable:v1.15.1`

also, IIRC, @nalind from the buildah team suggested 1.16 for even more improvements ... I would consider that change as well @vdemeester 

And moving forward, I think we should engage with him directly on these types of questions, which I've attempted to do here of course :-) ... please consider this the "formal introduction" @vdemeester @chmouel .... I'll mention this to him in the next meeting I'm in with him as well.

nalind replied:
1.16 should handle ADD and COPY better (faster) than it used to, and we've made correctness fixes as we've gone along.

Keep in mind that most of the gains we're seeing with vfs have happened because it tries to take advantage of the FICLONE and FICLONERANGE ioctls when copying the contents of one layer to create the basis for another layer, to which it then applies a layer diff to finish initializing the new layer's contents.  That ioctl() call will only succeed if the underlying filesystem supports it.  For OpenShift nodes, we can generally expect it to be XFS formatted with "-m reflink=1", which provides that ability, but I can't make guarantees about clusters in general.

gabemontero replied:
Ah .... very informative / useful @nalind  - thanks

so @vdemeester @chmouel @sbose78 @siamaksade .... vs. how we went with downstream openshift pipelines, perhaps we want to consider exposing the "performance vs. security" nuances around buildah in our upstream examples (as the performance impact is possibly more likely on non-openshift nodes)?

@sbose78 I know this was on your radar in particular recently ... thoughts?

chmouel replied:
thanks @nalind and @gabemontero, yeah maybe a documentation update mentionning nalind comment would be great for our users,

sbose78 replied:
I'm good as long as the default is vfs. If an advanced user wishes to flip that to "overlay", I would expect the advanced user to read documentation where the nuances (incl need for higher privileges) would be explained. :)

vdemeester commented:
:thinking: 
/hold

chmouel commented:
/retest

chmouel commented:
/lgtm

diff for people who needs it https://paste.gg/p/anonymous/e355c7d360f547ef9b5b309dfc46c8b6

chmouel commented:
test error seems legit : 

```
STEP 2: COMMIT registry:5000/nocode
level=error msg="Error while applying layer: ApplyLayer exit status 1 stdout:  stderr: remount /, flags: 0x44000: permission denied"
error committing base container: error copying layers and metadata for container "191af5fd2f25b48cfb087f36e88a01877df816ab4998ef3eed29b6018bd22f9e": Error committing the finished image: error adding layer with blob "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef": ApplyLayer exit status 1 stdout:  stderr: remount /, flags: 0x44000: permission denied
```

/hold

vinamra28 commented:
/test .*

vinamra28 commented:
/test .*

chmouel commented:
@vinamra28 no need to serial test, i think the CI error is legit error... not a fluke..

vinamra28 commented:
> @vinamra28 no need to serial test, i think the CI error is a legit error... not a fluke..

@chmouel I was just running the catlin CI :sweat_smile: 
`/test .*` runs only the catlin job

chmouel commented:
Ah cool cool thanks 👍

vinamra28 commented:
/test pull-catalog-catlin-lint

chmouel commented:
Related : 

https://github.com/containers/buildah/issues/2554#issuecomment-790124388

mdraijer commented:
Is there any progress on this issue? It is quite a problem running privileged containers just because Buildah needs them.
A solution was proposed a long time ago. This might be driving people towards other tools, such as Kaniko, that do not have this issue.

vdemeester commented:
/remove-lifecycle rotten

vdemeester commented:
> Is there any progress on this issue? It is quite a problem running privileged containers just because Buildah needs them.
> A solution was proposed a long time ago. This might be driving people towards other tools, such as Kaniko, that do not have this issue.

Updated, but as for kaniko, this still needs to run as root, so I've added the `securityContext` for running as root to be explicit.

vdemeester commented:
/close
Someone will need to carry this I think.
